Storage/OAuthStorage.php

```diff
@@ -164,12 +170,22 @@ public function checkUserCredentials(IOAuth2Client $client, $username, $password
         }
 
         if (null !== $user) {
-            $encoder = $this->encoderFactory->getEncoder($user);
-
-            if ($encoder->isPasswordValid($user->getPassword(), $password, $user->getSalt())) {
-                return array(
-                    'data' => $user,
-                );
+            $data = array(
+                'data' => $user,
+            );
+
+            if ($this->tokenPathAuth === 'plain') {
+                $encoder = $this->encoderFactory->getEncoder($user);
+
+                if ($encoder->isPasswordValid($user->getPassword(), $password, $user->getSalt())) {
+                    return $data;
+                }
+            } elseif ($this->tokenPathAuth === 'hash') {
+                if ($user->getPassword() === $password) {
```

Comment:
Its common to retreive the salt from the server with an end point to hash the plain password on the client and send it hashed to the server. The credentials indeed can be less secure but hashed there is no way a middle man can cash it with an sniffer